File: pyaerial/src/aerial/phy5g/ldpc/derate_match.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""pyAerial library - LDPC derate matching."""
from typing import Generic
from typing import List
from typing import Optional

import cupy as cp  # type: ignore
import numpy as np

from aerial import pycuphy  # type: ignore
from aerial.util.cuda import CudaStream
from aerial.phy5g.api import Array
from aerial.phy5g.config import PuschConfig


class LdpcDeRateMatch(Generic[Array]):
    """LDPC derate matching."""

    def __init__(self,
                 enable_scrambling: bool = True,
                 cuda_stream: Optional[CudaStream] = None) -> None:
        """Initialize LdpcDeRateMatch.

        Initialization does all the necessary memory allocations for cuPHY.

        Args:
            enable_scrambling (bool): Whether to descramble the bits before derate matching.
                Default: True.
            cuda_stream (Optional[CudaStream]): CUDA stream. If not given, a new CudaStream is
                created. Use ``with stream:`` to scope work; call ``stream.synchronize()``
                explicitly when sync is needed.
        """
        self._cuda_stream = CudaStream() if cuda_stream is None else cuda_stream

        # Create the cuPHY LDPC derate match object.
        self.pycuphy_ldpc_derate_match = pycuphy.LdpcDerateMatch(
            enable_scrambling,
            self._cuda_stream.handle
        )

    def derate_match(  # pylint: disable=too-many-arguments
            self,
            *,
            input_llrs: List[Array],
            pusch_configs: List[PuschConfig] = None,
            tb_sizes: List[int] = None,
            code_rates: List[float] = None,
            rate_match_lengths: List[int] = None,
            mod_orders: List[int] = None,
            num_layers: List[int] = None,
            redundancy_versions: List[int] = None,
            ndis: List[int] = None,
            cinits: List[int] = None,
            ue_grp_idx: List[int] = None) -> List[Array]:
        """LDPC derate matching function.

        Args:
            input_llrs (List[Array]): Input LLRs as a N x 1 array with dtype
                `np.float32`, where N is the number of LLRs coming from the equalizer. Ordering
                of this input data is `bitsPerQam x numLayers x numSubcarriers x numDataSymbols`.
                One entry per UE group.
            pusch_configs (List[PuschConfig]): List of PUSCH configuration objects, one per UE
                group. If this argument is given, the rest are ignored. If not given, all other
                arguments need to be given.
            tb_sizes (List[int]): Transport block sizes in bits without CRC, per UE.
            code_rates (List[float]): Code rates per UE.
            rate_match_lengths (List[int]): Number of rate matching output bits, the same as N,
                per UE.
            mod_orders (List[int]): Modulation order per UE.
            num_layers (List[int]): Number of layers per UE.
            redundancy_versions (List[int]): Redundancy version, i.e. 0, 1, 2, or 3, per UE.
            ndis (List[int]): New data indicator per UE.
            cinits (List[int]): The `c_init` value used for initializing scrambling for each UE.
            ue_grp_idx (List[int]): The UE group index for each UE. Default is one-to-one mapping.

        Returns:
            List[Array]: Derate matched LLRs for each UE.
        """
        cpu_copy = isinstance(input_llrs[0], np.ndarray)
        with self._cuda_stream:
            input_llrs = [cp.array(llr, order='F', dtype=cp.float16) for llr in input_llrs]

        # If PuschConfigs given, read the other parameters from that (the rest are ignored).
        if pusch_configs is not None:
            tb_sizes = []
            code_rates = []
            mod_orders = []
            num_layers = []
            redundancy_versions = []
            ndis = []
            cinits = []
            rate_match_lengths = []
            ue_grp_idx = []
            for idx, pusch_config in enumerate(pusch_configs):
                tb_sizes += [ue_config.tb_size * 8 for ue_config in pusch_config.ue_configs]
                code_rates += [ue_config.code_rate / 10240.
                               for ue_config in pusch_config.ue_configs]
                mod_orders += [ue_config.mod_order for ue_config in pusch_config.ue_configs]
                num_layers += [ue_config.layers for ue_config in pusch_config.ue_configs]
                redundancy_versions += [ue_config.rv for ue_config in pusch_config.ue_configs]
                ndis += [ue_config.ndi for ue_config in pusch_config.ue_configs]
                cinits += [
                    (ue_config.rnti << 15) + ue_config.data_scid
                    for ue_config in pusch_config.ue_configs
                ]

                num_data_sym = (np.array(pusch_config.dmrs_syms[
                    pusch_config.start_sym:pusch_config.start_sym + pusch_config.num_symbols
                ]) == 0).sum()
                num_ues = len(pusch_config.ue_configs)
                rate_match_lengths += [
                    num_data_sym * pusch_config.ue_configs[ue].mod_order * pusch_config.num_prbs *
                    12 * pusch_config.ue_configs[ue].layers
                    for ue in range(num_ues)
                ]
                ue_grp_idx += [idx for _ in range(num_ues)]

        # Make sure all parameters are given in this case.
        else:
            if tb_sizes is None:
                raise ValueError("Argument tb_sizes is not set!")
            if code_rates is None:
                raise ValueError("Argument code_rates is not set!")
            if rate_match_lengths is None:
                raise ValueError("Argument rate_match_lengths is not set!")
            if mod_orders is None:
                raise ValueError("Argument mod_orders is not set!")
            if num_layers is None:
                raise ValueError("Argument num_layers is not set!")
            if redundancy_versions is None:
                raise ValueError("Argument redundancy_versions is not set!")
            if ndis is None:
                raise ValueError("Argument ndis is not set!")
            if cinits is None:
                raise ValueError("Argument cinits is not set!")
            if ue_grp_idx is None:
                num_ues = len(tb_sizes)
                ue_grp_idx = list(range(num_ues))  # Default is one-to-one mapping.

        # Arrange the input data as cuPHY wants it.
        derate_match_inputs = []
        for idx, input_llr in enumerate(input_llrs):
            # Find first UE corresponding to this UE group.
            ue_idx = next(i for (i, v) in enumerate(ue_grp_idx) if idx == v)

            assert rate_match_lengths[ue_idx] <= (4 * 273 * 12 * 14 * 8), \
                "Maximum rate matching length exceeded!"

            if input_llr.ndim in (1, 2) or (input_llr.ndim == 4 and
                                            input_llr.shape[0] == mod_orders[ue_idx]):
                if input_llr.ndim == 2 and input_llr.shape[1] > 1:
                    raise ValueError("Invalid input data dimensions!")

                # The cuPHY derate matcher pads all inputs to 256-QAM as it supports MU-MIMO, and
                # different users may be different modulation order. We add this padding here.
                with self._cuda_stream:
                    derate_match_input = \
                        cp.zeros(
                            (8, int(rate_match_lengths[ue_idx] / mod_orders[ue_idx])),
                            dtype=cp.float16, order='F'
                        )
                    derate_match_input[:mod_orders[ue_idx], :] = \
                        input_llr.reshape((mod_orders[ue_idx], -1), order='F')

            elif input_llr.ndim != 4 or input_llr.shape[0] != 8:
                raise ValueError("Invalid input data dimensions!")

            else:
                derate_match_input = input_llr

            derate_match_inputs.append(derate_match_input)

        # Wrap CuPy arrays into pycuphy types.
        derate_match_inputs = [pycuphy.CudaArrayHalf(llr) for llr in derate_match_inputs]

        derate_matched_llrs = self.pycuphy_ldpc_derate_match.derate_match(
            derate_match_inputs,
            [np.uint32(tb_size) for tb_size in tb_sizes],
            [np.float32(code_rate) for code_rate in code_rates],
            [np.uint32(rate_match_len) for rate_match_len in rate_match_lengths],
            [np.uint8(mod_order) for mod_order in mod_orders],
            [np.uint8(nl) for nl in num_layers],
            [np.uint32(rv) for rv in redundancy_versions],
            [np.uint32(ndi) for ndi in ndis],
            [np.uint32(cinit) for cinit in cinits],
            [np.uint32(idx) for idx in ue_grp_idx]
        )

        with self._cuda_stream:
            derate_matched_llrs = [cp.array(elem) for elem in derate_matched_llrs]
            if cpu_copy:
                derate_matched_llrs = [elem.get(order='F') for elem in derate_matched_llrs]

        return derate_matched_llrs

File: 5GModel/nr_matlab/pxsch/derate_match.m
% SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
% SPDX-License-Identifier: Apache-2.0
%
% Licensed under the Apache License, Version 2.0 (the "License");
% you may not use this file except in compliance with the License.
% You may obtain a copy of the License at
%
% http://www.apache.org/licenses/LICENSE-2.0
%
% Unless required by applicable law or agreed to in writing, software
% distributed under the License is distributed on an "AS IS" BASIS,
% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
% See the License for the specific language governing permissions and
% limitations under the License.

function [derateCbs,nV_parity,derateCbsIndices,derateCbsIndicesSizes] = derate_match(LLR_demap, BGN, C, qam, ...
    K, F, K_prime, Zc, nl, N_data, rvIdx, Nref,G)

%Function undos rate matching. Takes LLRs of recieved bits and organizes
%them into LLRs of codeblocks.

%follows TS 38.212 section 5.4.2

%inputs:
%LLR_demap    --> demaped and descrambled LLRs

%outputs:
%deRateCbs    --> derate matched codeblocks
%nV_parity    --> number of parity nodes

%%
%PARAMATERS

%coding paramaters:
% BGN = PuschCfg.coding.BGN;         % 1 or 2. Indicates which base graph used
% C  = PuschCfg.coding.C;            % number of codeblocks
% qam = PuschCfg.coding.qam;         % bits per qam
% K = PuschCfg.coding.K;             % number of systematic bits per codeblock
% F = PuschCfg.coding.F;             % number of filler bits per codeblock
% K_prime = PuschCfg.coding.K_prime; %number of CRC encoded bits per codeblock (no filler)
% Zc = PuschCfg.coding.Zc;           % lifting size
%
% %allocation paramaters:
% nl = PuschCfg.mimo.nl;             % number of layers transmited by user
% N_data = PuschCfg.alloc.N_data;    % number of TF data resources in allocation

%%
%SIZE

%number of bits to be transmitted:
% G = N_data * qam * nl;

%derive number of rate matched bits per codeblock:
E = zeros(C,1);

for r = 0 : (C - 1)
    if r <= (C - mod( G / (nl * qam) , C) - 1)
        E(r + 1) = nl * qam * floor( G / (C * nl * qam) );
    else
        E(r + 1) = nl * qam * ceil( G / (C * nl * qam) );
    end
end


% apply LBMR
if BGN == 1
    N = Zc*66;
else
    N = Zc*50;
end


if Nref > 0
    N_cb = min(N, Nref);
else
    N_cb = N;
end


%%
%START

derateCbs = zeros(N,C);
derateCbsIndices = zeros(N,C);
derateCbsIndicesSizes = zeros(C,1);
current_bit = 0;

if BGN == 1
    switch rvIdx
        case 0
            k0 = 0;
        case 1
            k0 = floor(17*N_cb/(66*Zc))*Zc;
        case 2
            k0 = floor(33*N_cb/(66*Zc))*Zc;
        case 3
            k0 = floor(56*N_cb/(66*Zc))*Zc;
        otherwise
            error('rv is not supported...\n');
    end
elseif BGN == 2
    switch rvIdx
        case 0
            k0 = 0;
        case 1
            k0 = floor(13*N_cb/(50*Zc))*Zc;
        case 2
            k0 = floor(25*N_cb/(50*Zc))*Zc;
        case 3
            k0 = floor(43*N_cb/(50*Zc))*Zc;
        otherwise
            error('rv is not supported...\n');
    end
else
    error('BGN is not supported...\n');
end

%number of variable nodes:
if(BGN == 1)
    maxLLRPerCb    = Zc * 66;
    maxParityNodes = 46;
else
    maxLLRPerCb    = Zc * 50;
    maxParityNodes = 42;
end
nLLRPerCb    = min(floor(G / C) + k0, maxLLRPerCb);
nSymLLRPerCb = K - F - 2*Zc;
nParLLRPerCb = nLLRPerCb - nSymLLRPerCb;
nV_parity    = max(4, min(ceil(nParLLRPerCb / Zc), maxParityNodes));


for c = 1 : C

    %recover llrs corresponding to this codeblock:
    LLR_c = LLR_demap(current_bit + 1 : sum(E(1 : c)));

    %de-interleave:
    LLR_c = reshape(LLR_c, qam, E(c) / qam);
    LLR_c = LLR_c';
    LLR_c = LLR_c(:);

    %undo bit-selection:
    k = 0;
    j = 0;

    %display(['Zc=',num2str(Zc),' BGN=',num2str(BGN),' K=',num2str(K),' K_prime=',num2str(K_prime)]);
    %if (c == 1) || (c == C)
    %    global SimCtrl;
    %    display(['    {',num2str(E(c)),', ',num2str(K-2*Zc),', ',num2str(F),', ',num2str(k0),', ',num2str(N_cb),', ',num2str(Zc),'}, // ',SimCtrl.genTV.TVname,' c=',num2str(c)]);
    %end

    while k < E(c)
        %avoid filler bits:
        idx = mod(k0+j,N_cb) + 1;
        if ( idx <= (K_prime - 2*Zc) ) || ( idx > (K - 2*Zc) )
            cur_llr = derateCbs(idx,c) + LLR_c(k + 1);
            % clamp LLR as in cuPHY rate_matching.cu
            if (cur_llr > 10000)
                cur_llr = 10000;
            elseif (cur_llr < -10000)
                cur_llr = -10000;
            end
            derateCbs(idx,c) = cur_llr;
            derateCbsIndices(k+1,c) = idx;
            %line = sprintf('c,outIdx,inIdx,llr,%05d,%05d,%05d,%f',c,idx,k,derateCbs(idx,c));
            %display([' ',line]);
            k = k + 1;
        end
        j = j + 1;
    end

    derateCbsIndicesSizes(c) = E(c);

    %update current bit:
    current_bit = current_bit + E(c);
end

%add filler bits:
derateCbs(K_prime - 2*Zc + 1 : (K - 2*Zc),:) = 10000;
%for c = 1:size(derateCbs,2)
%    for idx = K_prime - 2*Zc + 1 : (K - 2*Zc)
%        line = sprintf('c,outIdx,inIdx,llr,%05d,%05d,%05d,%f',c,idx,99998,derateCbs(idx,c));
%        display([' ',line]);
%    end
%end

% threshold LLRs to match cuPHY
derateCbs(derateCbs > 10000)  = 10000;
derateCbs(derateCbs < -10000) = -10000;

% Adjust for Matlab to C indexing, includes padding used in C
derateCbsIndices = derateCbsIndices + 2*Zc - 1;



end
